src/Symfony/Component/Messenger/Adapter/PhpAmqp/Connection.php

```diff
@@ -0,0 +1,207 @@
+ *
+ * @author Samuel Roze <samuel.roze@gmail.com>
+ */
+class Connection
+{
+    private $connectionCredentials;
+    private $exchangeConfiguration;
+    private $queueConfiguration;
+    private $debug;
+
+    /**
+     * @var \AMQPChannel|null
+     */
+    private $amqpChannel;
+
+    /**
+     * @var \AMQPExchange|null
+     */
+    private $amqpExchange;
+
+    /**
+     * @var \AMQPQueue|null
+     */
+    private $amqpQueue;
+
+    public function __construct(array $connectionCredentials, array $exchangeConfiguration, array $queueConfiguration, bool $debug = false)
+    {
+        $this->connectionCredentials = $connectionCredentials;
+        $this->debug = $debug;
+        $this->exchangeConfiguration = $exchangeConfiguration;
+        $this->queueConfiguration = $queueConfiguration;
+    }
+
+    public static function fromDsn(string $dsn, bool $debug = false): self
+    {
+        if (false === $parsedUrl = parse_url($dsn)) {
+            throw new \InvalidArgumentException(sprintf('The given AMQP DSN "%s" is invalid.', $dsn));
+        }
+
+        $pathParts = isset($parsedUrl['path']) ? explode('/', trim($parsedUrl['path'], '/')) : array();
+        $amqpOptions = array(
+            'host' => $parsedUrl['host'] ?? 'localhost',
+            'port' => $parsedUrl['port'] ?? 5672,
+            'vhost' => isset($pathParts[0]) ? urldecode($pathParts[0]) : '/',
+            'queue' => array(
+                'name' => $queueName = $pathParts[1] ?? 'messages',
+            ),
+            'exchange' => array(
+                'name' => $queueName,
+            ),
+        );
+
+        if (isset($parsedUrl['user'])) {
+            $amqpOptions['login'] = $parsedUrl['user'];
+        }
+
+        if (isset($parsedUrl['pass'])) {
+            $amqpOptions['password'] = $parsedUrl['pass'];
+        }
+
+        if (isset($parsedUrl['query'])) {
+            parse_str($parsedUrl['query'], $parsedQuery);
+
+            $amqpOptions = array_merge($amqpOptions, $parsedQuery);
+        }
+
+        $exchangeOptions = $amqpOptions['exchange'];
+        $queueOptions = $amqpOptions['queue'];
+
+        unset($amqpOptions['queue']);
+        unset($amqpOptions['exchange']);
+
+        return new self($amqpOptions, $exchangeOptions, $queueOptions, $debug);
+    }
+
+    /**
+     * @throws \AMQPException
+     */
+    public function publish(string $body, array $headers = array())
+    {
+        if ($this->debug) {
+            $this->setup();
+        }
+
+        $this->exchange()->publish($body, null, AMQP_NOPARAM, array('headers' => $headers));
+    }
+
+    /**
+     * Waits and gets a message from the configured queue.
+     *
+     * @throws \AMQPException
+     */
+    public function waitAndGet(): ?\AMQPEnvelope
+    {
+        if ($this->debug) {
+            $this->setup();
+        }
+
+        $message = null;
+
+        try {
+            $this->queue()->consume(function (\AMQPEnvelope $envelope) use (&$message): bool {
```

Comment:
The callback might get a message from another queue (if several subscribed in one process).

Reply:
The amqp-ext is really fast though its consume method has one disadvantage it does not react on system signal, so if you have a big (non zero ) read timeout and getting signals they are ignored. bunny and php-amqplib throws an exception in this case.

Reply:
@sroze Have you tried to run consume method several times. It might create new subscription on rabbitmq side. 

You can check that by comparing consumer tags or by looking at queue consumer in Rabbit Managment UI.

Reply:
Thank you for your review @makasim. 

> The callback might get a message from another queue (if several subscribed in one process).

We don't allow this within the current implementation and wouldn't suggest we do for the long term within this AMQP bridge anyway. I guess this means we don't have to take this into account here.

> The amqp-ext is really fast though its consume method has one disadvantage it does not react on system signal

Might it be different with its version 1.9.x? I've added an integration test to ensure the signals are properly listened to.

Reply:
My point was that because we don't allow (yet?) to subscribe from the different queues in the same process, this should be OK.

Though, if we do, what would be your recommendation? In your experience, what's the best way to handle such situation?

Reply:
> My point was that because we don't allow (yet?) to subscribe from the different queues in the same process, this should be OK.

You cannot guaranty that, and a developer might take the AMQP channel and subscribe themselves or configure messenger the way that makes it possible. 

Is there a way to change the queue used? 

Once it happens you are in trouble. The behavior is really tricky and hard to reproduce, happens from time to time. I've spent a couple of months to fix it.

Reply:
I don't really like this design either.

1. It can work with only one queue at a time
2. Can not react (nicely) with signals
3. The PHP process can not does something else between 2 AMQP messages
4. It's blocking

Reply:
> It can work with only one queue at a time

Per running process, yes. 

> The PHP process can not does something else between 2 AMQP messages

What would it do? :)

> Can not react (nicely) with signals

It does stop to consume when receiving the signal.

Though, any alternative suggestion? 😃

Reply:
> It does stop to consume when receiving the signal.

It is not true for non zero timeout.

Reply:
Fair enough. So should we force a zero-timeout or do you guys see another solution?

Reply:
Do you have some kind of extensions or plugins that should be executed from time to time disregarding there is a message or not?

Reply:
> What would it do? :)

For exemple, In a specific app I worked on, the worker looks at the filesystem to check if something changed.

> Though, any alternative suggestion?

We use to use the `get` method instead of the `consume` method.

> Do you have some kind of extensions or plugins that should be executed from time to time disregarding there is a message or not?

1. Monitoring: send a ping to consul (for exemple) to said "I'm alive"
2. Check the memory of the current script. If > 100Mb (for exemple), commit suicide

Reply:
> We use to use the get method instead of the consume method.

In this context, I'm not sure to understand what would this change? Because as a developer, you wouldn't have access to this call either way, because this `waitAndGet` method is called inside the `AmqpReceiver`. If I understand your point correctly, this would mean that you'd like the `AmqpReceiver` to sometimes yield `null`s so you can do something else in between by creating a decorator, right? 🤔

Reply:
That's an idea. 

my point is: a consumer (cli process) will be very coupled with the underlying fetcher. 
It's not a daemon where we can do anything.

Reply:
After reflexion, that's a good point here. What if we simply `yield` `null`s [in the `AmqpReceiver` instead of `continue`ing](https://github.com/symfony/symfony/pull/26632/files#diff-da5657a5035db784d7c45b1b168d65fdR45) ? That use in combination with a timeout (so it actually yield nulls) and you're sorted out by simply creating a `ReceiverInterface` decorator, right?